JSON.mqh

```diff
@@ -90,24 +90,298 @@ class JSON {
       }
     }
 
-    return output + "\"";
+    return output + (includeQuotes ? "\"" : "");
   }
 
-  static string Stringify(float value, bool includeQuotes = false) {
+  static string ValueToString(float value, bool includeQuotes = false) {
     return (includeQuotes ? "\"" : "") + StringFormat("%.6f", value) + (includeQuotes ? "\"" : "");
   }
 
-  static string Stringify(double value, bool includeQuotes = false) {
+  static string ValueToString(double value, bool includeQuotes = false) {
     return (includeQuotes ? "\"" : "") + StringFormat("%.8f", value) + (includeQuotes ? "\"" : "");
   }
 
-  static string Stringify(Object* _obj, bool includeQuotes = false) {
+  static string ValueToString(Object* _obj, bool includeQuotes = false) {
     return (includeQuotes ? "\"" : "") + ((Object*)_obj).ToString() + (includeQuotes ? "\"" : "");
   }
   template <typename T>
-  static string Stringify(T value, bool includeQuotes = false) {
+  static string ValueToString(T value, bool includeQuotes = false) {
     return StringFormat("%s%s%s", (includeQuotes ? "\"" : ""), value, (includeQuotes ? "\"" : ""));
   }
+  
+  template<typename X>
+  static string Stringify(X& obj, bool trimWhitespace = false, int indentSize = 2) {
+    JsonSerializer serializer(NULL, JsonSerialize);
+    serializer.PassStruct(obj, "", obj);
+    
+    if (serializer.GetRoot()) {
+      return serializer.GetRoot().ToString(trimWhitespace, indentSize);
+    }
+    
+    // Error occured.
+    return "{\"error\": \"Cannot stringify object!\"}";
+  }
+
+  template<typename X>
+  static JsonNode* Parse(string data, X& obj) {
+    JsonNode* node = Parse(data);
+    
+    if (!node) {
+      // Parsing failed.
+      return NULL;
+    }
+
+    JsonSerializer serializer(node, JsonUnserialize);
+
+    // We don't use result. We parse data as it is.
+    obj.Serialize(serializer);
+
+    return node;
+  }
+
+  static JsonNode* Parse(string data) {
+    JsonNodeType type;
+    if (StringGetCharacter(data, 0) == '{')
+      type = JsonNodeObject;
+    else
+    if (StringGetCharacter(data, 0) == '[')
+      type = JsonNodeArray;
+    else {
+      return GracefulReturn("Failed to parse JSON. It must start with either \"{\" or \"[\".", 0, NULL, NULL);
+    }
+
+    JsonNode* root = NULL;
+    JsonNode* current = NULL;
+    JsonNode* node = NULL;
+    
+    string extracted;
+    
+    bool isOuterScope = true;
+    bool expectingKey = false;
+    bool expectingValue = false;
+    bool expectingSemicolon = false;
+    JsonParam* key = NULL;
+    JsonParam* value = NULL;
+    
+    for (unsigned int i = 0; i < (unsigned int)StringLen(data); ++i)
+    {
+    #ifdef __MQL5__
+      unsigned short ch = StringGetCharacter(data, i);
+    #else
+      unsigned short ch = StringGetChar(data, i);
```

Comment:
It seems MT4 also supports [StringGetCharacter()](https://docs.mql4.com/strings/stringgetcharacter).